flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriterFactory.java

```diff
@@ -69,8 +71,16 @@ public RowDataTaskWriterFactory(Table table,
 
     if (equalityFieldIds == null || equalityFieldIds.isEmpty()) {
       this.appenderFactory = new FlinkAppenderFactory(schema, flinkSchema, table.properties(), spec);
+    } else if (upsert) {
+      // In upsert mode, only the new row is emitted using INSERT row kind. Therefore, the deletion schema
+      // must be only the equality fields for that the writer can use the new row to issue a delete by key.
+      //
+      // Outside the upsert case, the entire deleted row is emitted as UPDATE_BEFORE or DELETE row kind,
+      // so the deletion schema must use the whole rows schema in case elements are deleted based on additional
+      // fields.
```

Comment:
Nit: I don't think this part of the comment is needed, but it's okay to have. I would move it to the next block if you want to include it so that it applies directly to the code that implements it.

Reply:
I can move it.

Reply:
Removed it.